File: Part1_강의자료/Ch07_이분탐색/문제별코드/6236_용돈관리/6236.py
import sys
input = sys.stdin.readline

N, M = map(int, input().split())
use_amounts = [int(input()) for _ in range(N)]

def is_possible(mid):
    draw_cnt, current_amount = 1, mid
    for use_amount in use_amounts:
        current_amount -= use_amount
        if current_amount < 0:
            current_amount = mid - use_amount
            draw_cnt += 1
    return draw_cnt <= M

l, r = max(use_amounts), sum(use_amounts)
while l < r:
    mid = (l + r) // 2
    if is_possible(mid):
        r = mid
    else:
        l = mid + 1
print(l)

File: Part1_강의자료/Ch07_이분탐색/문제별코드/6236_용돈관리/6236.cpp
#include <algorithm>
#include <stdio.h>
#include <vector>
using namespace std;

bool is_possible(vector<int> &use_amounts, int M, int mid)
{
    int draw_cnt = 1, current_amount = mid;
    for (auto & use_amount : use_amounts)
    {
        current_amount -= use_amount;
        if (current_amount < 0)
        {
            current_amount = mid - use_amount;
            draw_cnt++;
        }
    }
    return draw_cnt <= M;
}

int main()
{
    int N, M;
    scanf("%d %d", &N, &M);
    vector<int> use_amounts(N);
    int l = -1, r = 0;
    
    for (int i = 0; i < N; i++)
    {
        scanf("%d", &use_amounts[i]);
        l = max(l, use_amounts[i]);
        r += use_amounts[i];
    }

    while (l < r)
    {
        int mid = (l + r) / 2;
        if (is_possible(use_amounts, M, mid))
            r = mid;
        else
            l = mid + 1;
    }
    printf("%d\n", l);
    return 0;
}
